aws/data_source_aws_region.go

```diff
@@ -36,48 +35,74 @@ func dataSourceAwsRegion() *schema.Resource {
 }
 
 func dataSourceAwsRegionRead(d *schema.ResourceData, meta interface{}) error {
-	conn := meta.(*AWSClient).ec2conn
 	currentRegion := meta.(*AWSClient).region
 
-	req := &ec2.DescribeRegionsInput{}
+	var matchingRegion *endpoints.Region
 
-	req.RegionNames = make([]*string, 0, 2)
-	if name := d.Get("name").(string); name != "" {
-		req.RegionNames = append(req.RegionNames, aws.String(name))
+	if v, ok := d.GetOk("endpoint"); ok {
+		endpoint := v.(string)
+		for _, partition := range endpoints.DefaultPartitions() {
+			for _, region := range partition.Regions() {
+				regionEndpointEc2, err := region.ResolveEndpoint(endpoints.Ec2ServiceID)
+				if err != nil {
+					return err
+				}
+				if strings.TrimPrefix(regionEndpointEc2.URL, "https://") == endpoint {
+					matchingRegion = &region
+					break
+				}
+			}
+		}
+		if matchingRegion == nil {
+			return fmt.Errorf("region not found for endpoint: %s", endpoint)
+		}
```

Comment:
It took me a while to understand what this block of code does. 😅  Would you mind extracting it into something like `matchingRegion, err := findRegionByEndpoint(endpoint)`?

Reply:
That's a great idea! Let me implement that.